### PR TITLE
implement the rmw_gid_t related functions

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/ros2#91

See: https://github.com/ros2/rmw/pull/35

This still needs to be tested. It seems to work, but I'm not 100% certain that the comparison is sufficient. I'm looking for a better wait to compare the OpenSplice GID's.

### CODE REVIEW

File: rmw_opensplice_cpp/src/functions.cpp [hunk 1]
```diff
@@ -1034,53 +1039,80 @@ rmw_destroy_subscription(rmw_node_t * node, rmw_subscription_t * subscription)
 rmw_ret_t
 rmw_take(const rmw_subscription_t * subscription, void * ros_message, bool * taken)
 {
+  auto publisher_gid = rmw_take_with_gid(subscription, ros_message, taken);
+  bool result = (publisher_gid != nullptr);
+  rmw_destroy_gid(publisher_gid);
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/functions.cpp]: I don't like the idea that every `rmw_take` creates a gid and destroys it immediately again.

Comment (wjwwood) [inline rmw_opensplice_cpp/src/functions.cpp]: I agree, but this was the shortest path. I'll refactor it into three functions, one that does the work but also returns the information needed to the create the publisher_gid, `take` which is normal, but doesn't create a gid, and take with gid which uses the extra info to make a gid too.

Comment (wjwwood) [inline rmw_opensplice_cpp/src/functions.cpp]: Ok, have a look at eecd08c.

File: rmw_opensplice_cpp/src/functions.cpp [hunk 2]
```diff
@@ -1083,6 +1090,40 @@ rmw_take(const rmw_subscription_t * subscription, void * ros_message, bool * tak
...
+  rmw_gid_t * sender_gid = static_cast<rmw_gid_t *>(rmw_allocate(sizeof(rmw_gid_t)));
+  if (!sender_gid) {
+    RMW_SET_ERROR_MSG("failed to allocate memory");
+    return nullptr;
+  }
+  OpenSplicePublisherGID * detail =
+    static_cast<OpenSplicePublisherGID *>(rmw_allocate(sizeof(OpenSplicePublisherGID)));
+  if (!detail) {
+    RMW_SET_ERROR_MSG("failed to allocate memory");
+    return nullptr;
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/functions.cpp]: Needs to free `sender_gid`.

File: rmw_opensplice_cpp/src/functions.cpp [hunk 3]
```diff
@@ -1083,6 +1090,40 @@ rmw_take(const rmw_subscription_t * subscription, void * ros_message, bool * tak
...
 }
 
+rmw_ret_t
+rmw_take(const rmw_subscription_t * subscription, void * ros_message, bool * taken)
+{
+  return _take(subscription, ros_message, taken, nullptr);
+}
+
+rmw_gid_t *
+rmw_take_with_gid(const rmw_subscription_t * subscription, void * ros_message, bool * taken)
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/functions.cpp]: How does this function signal the specific error code? Shouldn't it return `rmw_ret_t` and have an additional argument `rmw_gid_t * &`?

File: rmw_opensplice_cpp/src/functions.cpp [hunk 4]
```diff
@@ -2094,4 +2135,92 @@ rmw_count_subscribers(
...
+  detail->publisher_system_id = topic_writer_gid.systemId;
+  gid->implementation_identifier = opensplice_cpp_identifier;
+  gid->data = detail;
+  return gid;
+}
+
+void
+rmw_destroy_gid(rmw_gid_t * gid)
+{
+  if (gid) {
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/functions.cpp]: This must check the `implementation_identifier`.

**Comment by dirk-thomas** [comment]
The gid is one additional meta information we add in this PR. I expect that we might want to add another piece of information in the future. May be we should encapsulate the gid in a struct similar to the DDS `SampleInfo`.

**Comment by wjwwood** [comment]
I agree. I was thinking of making `rmw_publisher_info_t` and `rmw_message_info_t` or `rmw_message_instance_info_t` classes, but I think they would both contain a `rmw_gid_t`, so I avoided that for now as optional.

File: rmw_opensplice_cpp/src/functions.cpp (hunk 3 above)
Comment (wjwwood) [inline rmw_opensplice_cpp/src/functions.cpp]: It returns a nullptr to indicate an error and sets the error string. This is a pretty common thing across all of our functions. If we want to return an error specific error code we need to refactor many of our rmw functions. I don't want to do that here.

File: rmw_opensplice_cpp/src/functions.cpp (hunk 2 above)
Comment (wjwwood) [inline rmw_opensplice_cpp/src/functions.cpp]: Yup, fixed in 9e4bb2f.

File: rmw_opensplice_cpp/src/functions.cpp (hunk 4 above)
Comment (wjwwood) [inline rmw_opensplice_cpp/src/functions.cpp]: Fixed in b9db525 and ros2/rmw@be65bf6

**Comment by jacquelinekay** [comment]
+1

File: rmw_opensplice_cpp/src/functions.cpp [hunk 5]
```diff
@@ -1083,6 +1106,39 @@ rmw_take(const rmw_subscription_t * subscription, void * ros_message, bool * tak
...
+    return RMW_RET_ERROR;
+  }
+  DDS::InstanceHandle_t sending_publication_handle;
+  auto ret = _take(subscription, ros_message, taken, &sending_publication_handle);
+  if (ret != RMW_RET_OK) {
+    // Error string is already set.
+    return RMW_RET_ERROR;
+  }
+
+  rmw_gid_t * sender_gid = const_cast<rmw_gid_t *>(&message_info->publisher_gid);
```
Comment (dirk-thomas) [inline rmw_opensplice_cpp/src/functions.cpp]: Neither `message_info` nor `publisher_gid` seems to be const. Why is the const_cast necessary here?

**Comment by dirk-thomas** [comment]
LGTM

**Comment by tfoote** [comment]
+1

File: rmw_opensplice_cpp/src/functions.cpp (hunk 5 above)
Comment (wjwwood) [inline rmw_opensplice_cpp/src/functions.cpp]: Const was dropped in a refactoring. 615b37e

**Comment by wjwwood** [comment]
Squashed.

**Pull request merged**